mender-update/http_resumer/http_resumer_test.cpp

```diff
@@ -861,3 +861,217 @@ TEST_F(DownloadResumerTest, SmallIntervalsErrorOnFirstRead) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+
+			auto result = exp_req.value()->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetStatusCodeAndMessage(200, "OK, almost");
+			resp->SetHeader("Content-Length", "0");
+			resp->AsyncReply([](error::Error err) { ASSERT_EQ(error::NoError, err); });
+		});
+
+	// Request
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	req->SetAddress("http://127.0.0.1:" TEST_PORT);
+
+	// Client
+	http::ClientConfig client_config;
+	shared_ptr<http_resumer::DownloadResumerClient> client =
+		make_shared<http_resumer::DownloadResumerClient>(client_config, loop);
+	client->SetSmallestWaitInterval(chrono::seconds(2));
+
+	bool body_handler_called = false;
+
+	client->AsyncCall(
+		req,
+		[](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+			auto resp = exp_resp.value();
+
+			EXPECT_EQ(resp->GetStatusCode(), 200);
+			EXPECT_EQ(resp->GetStatusMessage(), "OK, almost");
+		},
+		[&loop, &body_handler_called](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_TRUE(exp_resp) << exp_resp.error().String();
+			auto resp = exp_resp.value();
+
+			EXPECT_EQ(resp->GetStatusCode(), 200);
+			EXPECT_EQ(resp->GetStatusMessage(), "OK, almost");
+
+			body_handler_called = true;
+
+			loop.Stop();
+		});
+
+	loop.Run();
+
+	EXPECT_TRUE(body_handler_called);
+}
+
+TEST_F(DownloadResumerTest, UserCancelInHeaderHandler) {
+	TestEventLoop loop(chrono::seconds(10));
+
+	// Server
+	http::ServerConfig server_config;
+	http::Server server(server_config, loop);
+
+	server.AsyncServeUrl(
+		"http://127.0.0.1:" TEST_PORT,
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+		},
+		[](http::ExpectedIncomingRequestPtr exp_req) {
+			ASSERT_TRUE(exp_req) << exp_req.error().String();
+
+			auto result = exp_req.value()->MakeResponse();
+			ASSERT_TRUE(result);
+			auto resp = result.value();
+
+			resp->SetHeader("Content-Length", to_string(RangeBodyOfXes::TARGET_BODY_SIZE));
+			resp->SetBodyReader(make_shared<RangeBodyOfXes>());
+			resp->SetStatusCodeAndMessage(200, "Success");
+			resp->AsyncReply([](error::Error err) { ASSERT_NE(error::NoError, err); });
+		});
+
+	// Request
+	auto req = make_shared<http::OutgoingRequest>();
+	req->SetMethod(http::Method::GET);
+	req->SetAddress("http://127.0.0.1:" TEST_PORT);
+
+	// Client
+	http::ClientConfig client_config;
+	shared_ptr<http_resumer::DownloadResumerClient> client =
+		make_shared<http_resumer::DownloadResumerClient>(client_config, loop);
+	client->SetSmallestWaitInterval(chrono::seconds(2));
+
+	client->AsyncCall(
+		req,
+		[&client](http::ExpectedIncomingResponsePtr exp_resp) { client->Cancel(); },
+		[](http::ExpectedIncomingResponsePtr exp_resp) {
+			ASSERT_FALSE(exp_resp);
+			EXPECT_EQ(exp_resp.error().code, make_error_condition(errc::operation_canceled));
```

Comment:
Can't you stop the loop here instead?

Reply:
This test I also copied (see above) and interestingly stopping from the loop doesn't work (which is a bug in itself). But it also doesn't work correctly on `http_test.cpp` so that we need to follow-up separately.

Reply:
Just to be clear. The following patch for `http_test.cpp` fails the test
```
$ git diff common/http_test.cpp
diff --git a/common/http_test.cpp b/common/http_test.cpp
index 4ba364ec..ccf1b3f6 100644
--- a/common/http_test.cpp
+++ b/common/http_test.cpp
@@ -915,13 +915,14 @@ TEST(HttpTest, TestClientCancelInHeaderHandler) {
        client.AsyncCall(
                req,
                [&client](http::ExpectedIncomingResponsePtr exp_resp) { client.Cancel(); },
-               [](http::ExpectedIncomingResponsePtr exp_resp) {
+               [&loop](http::ExpectedIncomingResponsePtr exp_resp) {
                        ASSERT_FALSE(exp_resp);
                        EXPECT_EQ(exp_resp.error().code, make_error_condition(errc::operation_canceled));
+                       loop.Stop();
                });
 
-       events::Timer timer(loop);
-       timer.AsyncWait(chrono::milliseconds(500), [&loop](error::Error err) { loop.Stop(); });
+       // events::Timer timer(loop);
+       // timer.AsyncWait(chrono::milliseconds(500), [&loop](error::Error err) { loop.Stop(); });
 
        loop.Run();
 }
```

Reply:
> This test I also copied (see above) and interestingly stopping from the loop doesn't work (which is a bug in itself). 

This I fixed in the previous force-push. The sleep is not part of this test anymore. Sorry for not being explicit.